ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,241 @@
+          BlobStoreStats.getContainerId(messageInfo.getStoreKey()), messageInfo.getSize());
+    }
+  }
+
+  /**
+   * Process a record in the index and populate the valid data size per log segment map.
+   * @param logSegmentValidSizeMap the valid data size per log segment map
+   * @param referenceTimeInMs the reference time used to decide whether or not a record is valid
+   * @param messageInfo the {@link MessageInfo} to be processed
+   * @param lastModifiedTimeInMs the time stamp indicating when the record that is being processed was added
+   * @param logSegmentName the log segment name of the record that is being processed
+   * @param backwardSearchSpan a {@link FileSpan} that spans from the beginning of the index to the beginning of the
+   *                           previous index segment and is used for looking up corresponding put record when applicable
+   * @throws StoreException
+   */
+  private void populateLogSegmentMap(Map<String, Long> logSegmentValidSizeMap, long referenceTimeInMs,
+      MessageInfo messageInfo, long lastModifiedTimeInMs, String logSegmentName, FileSpan backwardSearchSpan)
+      throws StoreException {
+
+    if (messageInfo.isDeleted()) {
+      // delete record
+      updateMapHelper(logSegmentValidSizeMap, logSegmentName, messageInfo.getSize());
+      if (lastModifiedTimeInMs < referenceTimeInMs) {
+        // delete that needs to be accounted for
+        IndexValue putIndexValue = index.findKey(messageInfo.getStoreKey(), backwardSearchSpan);
+        if (putIndexValue != null && !putIndexValue.isFlagSet(IndexValue.Flags.Delete_Index) && !isExpired(
+            putIndexValue.getExpiresAtMs(), referenceTimeInMs)) {
+          // correct previously counted value only if the following conditions are true:
+          // 1. the put and delete record are not in the same index segment (putIndexValue != null)
+          // 2. the found record is actually a put record (!putIndexValue.isFlagSet)
+          // 3. the found put record is not expired
+          updateMapHelper(logSegmentValidSizeMap, putIndexValue.getOffset().getName(), putIndexValue.getSize() * -1);
+        }
+      }
+    } else if (!isExpired(messageInfo.getExpirationTimeInMs(), referenceTimeInMs)) {
+      // put record that is not expired
+      updateMapHelper(logSegmentValidSizeMap, logSegmentName, messageInfo.getSize());
+    }
+  }
+
+  private boolean isExpired(long expirationTimeInMs, long referenceTimeInMs) {
+    return expirationTimeInMs < referenceTimeInMs && expirationTimeInMs != Utils.Infinite_Time;
+  }
+
+  /**
+   * Helper function to update nested map data structure.
+   * @param nestedMap nested {@link Map} to be updated
+   * @param firstKey of the nested map
+   * @param secondKey of the nested map
+   * @param value the value to be added at the corresponding entry
+   */
+  private void updateNestedMapHelper(Map<String, Map<String, Long>> nestedMap, String firstKey, String secondKey,
```

Comment:
I thought it's better to have function name to describe what it does rather than what is used for. In our case this function can update all nested maps and not just limited to ValidSizePerServiceId. In fact, in the bucketing implementation this is also called by the helper to update the buckets.